contracts/core/SequencerProofVerifier.sol

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright 2024, Polymer Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity 0.8.15;
+
+import {RLPReader} from "optimism/libraries/rlp/RLPReader.sol";
+import {RLPWriter} from "optimism/libraries/rlp/RLPWriter.sol";
+import {L1Header, Ics23Proof, OpIcs23Proof} from "../interfaces/IProofVerifier.sol";
+import {ISignatureVerifier} from "../interfaces/ISignatureVerifier.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/**
+ * @title OptimisticProofVerifier
+ * @notice Verifies proofs related to Optimistic Rollup state updates
+ * @author Polymer Labs
+ */
+contract SequencerSignatureVerifier is ISignatureVerifier {
+    using RLPReader for RLPReader.RLPItem;
+    using RLPReader for bytes;
+
+    // @notice known L2 Output Oracle contract address to verify state update proofs against
+    address public immutable sequencer;
+
+    error InvalidSequencerSignature();
+
+    constructor(address sequencer_) {
+        sequencer = sequencer_;
+    }
+
+    function verifyStateUpdate(L1Header calldata l1header, bytes32 appHash, uint256 l2Height, bytes calldata signature)
+        external
+        view
+    {
+        _verifySequencerSignature(
+            l2Height, appHash, keccak256(RLPWriter.writeList(l1header.header)), l1header.stateRoot, signature
+        );
+    }
+
+    /**
+     * @dev Prove that a given state is not part of a proof
+     * @dev this method is mainly used for packet timeouts, which is currently not implemented
+     */
+    function verifyNonMembership(bytes32, bytes calldata, Ics23Proof calldata) external pure {
+        revert MethodNotImplemented();
+    }
+
+    /**
+     * @inheritdoc ISignatureVerifier
+     * @dev verifies a chain of ICS23 proofs
+     * Each computed subroot starting from index 0 must match the value of the next proof (hence chained proofs).
+     * The cosmos SDK and ics23 support chained proofs to switch between different proof specs.
+     * Custom proof specs are not supported here. Only Iavl and Tendermint or similar proof specs are supported.
+     */
+    function verifyMembership(bytes32 appHash, bytes calldata key, bytes calldata value, Ics23Proof calldata proofs)
```

Comment:
Seems like this and the verifyNonMembership functions could be extracted into a library to be reused across all the proof verifiers?

Reply:
see[ this comment](https://github.com/open-ibc/vibc-core-smart-contracts/pull/199#discussion_r1765117411)